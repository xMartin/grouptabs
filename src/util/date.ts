/**
 * Parses a date string in the format `yyyy-MM-dd` as being used by `<input type="date">`
 * consistently across different hosts (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/parse).
 */
export function parseDate(input: string) {
  // backwards compatibility: strip time info from format being used earlier
  const date = input.substring(0, 10);

  const params = date.split("-").map((x) => parseInt(x, 10));

  if (params.length < 3) {
    throw new Error(
      `Invalid date format (should start with "yyyy-MM-dd"): ${input}`
    );
  }

  return new Date(params[0], params[1] - 1, params[2]);
}

/**
 * Formats a date to the format `yyyy-MM-dd` as being used by `<input type="date">`
 */
export function formatDate(date: Date) {
  const month = date.getMonth() + 1;
  const formattedMonth = month < 10 ? "0" + month : month;
  const day = date.getDate();
  const formattedDay = day < 10 ? "0" + day : day;
  return `${date.getFullYear()}-${formattedMonth}-${formattedDay}`;
}

export function formatHumanDate(date: Date | string) {
  if (typeof date === "string") {
    date = parseDate(date);
  }

  if (isToday(date)) {
    return "Today";
  }

  if (isYesterday(date)) {
    return "Yesterday";
  }

  return date.toLocaleDateString();
}

export function addDays(date: Date, days: number) {
  const ms = date.getTime();
  const daysInMs = days * 1000 * 60 * 60 * 24;
  return new Date(ms + daysInMs);
}

export function isSameDay(date1: Date, date2: Date) {
  return formatDate(date1) === formatDate(date2);
}

export function isToday(date: Date | string) {
  if (typeof date === "string") {
    date = parseDate(date);
  }

  const today = new Date();
  return isSameDay(today, date);
}

export function isYesterday(date: Date | string) {
  if (typeof date === "string") {
    date = parseDate(date);
  }

  const yesterday = addDays(new Date(), -1);
  return isSameDay(yesterday, date);
}
