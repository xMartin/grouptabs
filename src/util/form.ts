// Prevent warning about switching from uncoltrolled to controlled inputs
// https://github.com/facebook/react/issues/6222
export function control(value: any) /* TODO return type any but undefined */ {
  return value === undefined ? "" : value;
}
