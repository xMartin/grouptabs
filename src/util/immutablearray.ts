// immutable array helpers

export default class ImmutableArrayHelper {
  static addUniq<T>(array: T[], item: T): T[] {
    const index = array.indexOf(item);
    if (index !== -1) {
      return array;
    }

    return array.concat(item);
  }

  static remove<T>(array: T[], index: number): T[] {
    array = array.slice(); // copy
    array.splice(index, 1);
    return array;
  }

  static removeItem<T>(array: T[], item: T): T[] {
    const index = array.indexOf(item);

    if (index !== -1) {
      return ImmutableArrayHelper.remove(array, index);
    }

    return array;
  }
}
