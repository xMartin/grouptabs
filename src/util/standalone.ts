import { History } from "history";

// Safari on iOS has a "standalone" feature, aka "add to homescreen".
// Every time the user opens this app, it restarts from the URL
// that was originally set as the bookmark.
// The experience we want is that it starts from where the user navigated to last.
// In other similar environments, e.g. Chrome on Android add to homescreen
// the app is not always restarted when re-opened but will get killed after a while as well.

const isStandalone =
  !!(window.navigator as any).standalone ||
  window.matchMedia("(display-mode: standalone)").matches ||
  document.referrer.includes("android-app://");
const storageKey = "grouptabs_last_visited_location";

export function restoreLocation() {
  if (!isStandalone) {
    return;
  }

  let lastVisitedUrl = window.localStorage.getItem(storageKey);
  if (lastVisitedUrl) {
    // convert potentially legacy hash URL to clean URL
    lastVisitedUrl = lastVisitedUrl.replace("/#/tabs/", "/tabs/");
    window.history.replaceState(null, "", lastVisitedUrl);
  }
}

export function startPersistingLocation(history: History) {
  if (!isStandalone) {
    return;
  }

  history.listen(() =>
    window.localStorage.setItem(storageKey, window.location.href)
  );
}
