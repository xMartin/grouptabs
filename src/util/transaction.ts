import { Transaction, TransactionType } from "../types";

/**
 * Determines transaction type from transaction data.
 *
 * Old transaction data might not have `transactionType` specified as "DIRECT"
 * but semantically still be a direct transaction.
 */
function getTransactionType(transaction: Transaction): TransactionType {
  if (transaction.transactionType === TransactionType.DIRECT) {
    return transaction.transactionType;
  }

  // find out if it's a transaction with two participants where one paid amount X
  // and the other paid amount -X
  if (transaction.participants.length === 2) {
    const amounts = transaction.participants.map((participant) => {
      return participant.amount;
    });
    if (amounts[0] && amounts[0] === -amounts[1]) {
      return TransactionType.DIRECT;
    }
  }

  return TransactionType.SHARED;
}

export function mapTransaction(transaction: Transaction): Transaction {
  const actualTransactionType = getTransactionType(transaction);

  if (transaction.transactionType !== actualTransactionType) {
    return {
      ...transaction,
      transactionType: actualTransactionType,
    };
  }

  return transaction;
}
