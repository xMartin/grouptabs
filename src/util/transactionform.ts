import {
  TransactionFormState,
  Account,
  Transaction,
  TransactionFormParticipantInputType as InputType,
  TransactionFormParticipantStatus as Status,
  TransactionType,
  DocumentType,
  TransactionFormSharedState,
} from "../types";
import { v4 as uuidv4 } from "uuid";
import orderBy from "lodash/orderBy";
import { formatDate, parseDate } from "./date";
import { mapTransaction } from "./transaction";

export const NEW_PARTICIPANT_OPTION = "NEW_PARTICIPANT";

export function createNewParticipant() {
  return {
    id: uuidv4(),
    inputType: InputType.NEW,
    status: Status.JOINED,
  };
}

function createParticipantInputDataForEmptyTab(): TransactionFormState["shared"] {
  return [1, 2].map(() => createNewParticipant());
}

function createParticipantInputData(
  accounts: Account[],
  transaction?: Transaction,
): TransactionFormState["shared"] {
  if (!accounts.length) {
    return createParticipantInputDataForEmptyTab();
  }

  const participantInputs: TransactionFormState["shared"] = accounts.map(
    (account) => {
      let status;
      let amount;
      if (transaction) {
        const value = transaction.participants.find((participantValue) => {
          return participantValue.participant === account.participant;
        });
        status =
          value && value.amount > 0
            ? Status.PAID
            : value && value.amount === 0
              ? Status.JOINED
              : Status.NONE;
        if (value?.amount !== undefined && value.amount > 0) {
          amount = value.amount;
        }
      } else {
        // For exactly two people in the tab, always set JOINED
        status = accounts.length === 2 ? Status.JOINED : Status.NONE;
      }

      return {
        id: uuidv4(),
        inputType: InputType.EXISTING,
        participant: account.participant,
        status,
        amount,
      };
    },
  );

  // sort participants by 1) amount paid 2) joined 3) alphabetical
  return orderBy(
    participantInputs,
    ["status", "amount", "participant"],
    ["desc", "desc", "asc"],
  );
}

interface Inputs {
  from?: string;
  to?: string;
  amount?: number;
}

function transactionParticipants2Inputs(participants: Account[]): Inputs {
  const result: Inputs = {};

  for (const participant of participants) {
    if (!result.amount) {
      result.amount = Math.abs(participant.amount);
    }

    if (!result.from && participant.amount > 0) {
      result.from = participant.participant;
    }

    if (!result.to && participant.amount < 0) {
      result.to = participant.participant;
    }
  }

  return result;
}

function createDirectInputData(
  accounts: Account[],
  transaction?: Transaction,
): TransactionFormState["direct"] {
  const inputProps = transactionParticipants2Inputs(
    transaction?.participants || [],
  );

  const mostNegativeParticipant = accounts[0]?.participant;
  const mostPositiveParticipant = accounts[accounts.length - 1]?.participant;

  const tabParticipants = accounts.map((account) => account.participant);
  const options = tabParticipants.concat(NEW_PARTICIPANT_OPTION);

  return {
    options,
    from: inputProps.from || mostNegativeParticipant || NEW_PARTICIPANT_OPTION,
    to: inputProps.to || mostPositiveParticipant || NEW_PARTICIPANT_OPTION,
    amount: inputProps.amount,
  };
}

export function createFormData(
  accounts: Account[],
  transaction?: Transaction,
): TransactionFormState {
  transaction = transaction && mapTransaction(transaction);

  const shared = createParticipantInputData(
    accounts,
    transaction?.transactionType === TransactionType.SHARED
      ? transaction
      : undefined,
  );
  const direct = createDirectInputData(
    accounts,
    transaction?.transactionType === TransactionType.DIRECT
      ? transaction
      : undefined,
  );
  const form: TransactionFormState = {
    transactionType: TransactionType.SHARED,
    date: formatDate(new Date()),
    shared,
    direct,
  };

  if (transaction) {
    form.transactionType = transaction.transactionType;
    form.date = formatDate(parseDate(transaction.date));
    form.description = transaction.description;
  }

  return form;
}

function mapParticipant(
  participant: TransactionFormSharedState,
): Account | undefined {
  if (!participant.participant || participant.status === Status.NONE) {
    return;
  }

  // trim new participant names
  // don't trim existing ones to be backwards compatible
  const participantName =
    participant.inputType === InputType.NEW
      ? participant.participant.trim()
      : participant.participant;

  return {
    participant: participantName,
    amount:
      participant.status === Status.PAID && participant.amount
        ? participant.amount
        : 0,
  };
}

function mapParticipantsOfSharedTransaction(
  participants: TransactionFormState["shared"],
): Account[] {
  return participants
    .map(mapParticipant)
    .filter((participant) => participant) as Account[];
}

function mapParticipantsOfDirectTransaction(
  direct: TransactionFormState["direct"],
): Account[] {
  // trim new participant names
  // don't trim existing ones to be backwards compatible
  const from =
    direct.from === NEW_PARTICIPANT_OPTION
      ? direct.fromNew?.trim()
      : direct.from;
  const to =
    direct.to === NEW_PARTICIPANT_OPTION ? direct.toNew?.trim() : direct.to;

  const participants = [
    {
      participant: from,
      amount: direct.amount,
    },
    {
      participant: to,
      amount: -(direct.amount as number),
    },
  ] as Account[];

  return participants;
}

export function mapFormDataToTransaction(
  formData: TransactionFormState,
  tabId: string,
  id?: string,
): Transaction {
  let participants: Account[];
  if (formData.transactionType === TransactionType.SHARED) {
    participants = mapParticipantsOfSharedTransaction(formData.shared);
  } else {
    participants = mapParticipantsOfDirectTransaction(formData.direct);
  }

  return {
    id: id || uuidv4(),
    type: DocumentType.TRANSACTION,
    tabId,
    timestamp: new Date().toJSON(),
    transactionType: formData.transactionType,
    description: formData.description || "",
    date: formData.date,
    participants,
  };
}

function validateShared(participants: TransactionFormState["shared"]): boolean {
  const joinedParticipants = participants.filter(
    (participant) => participant.status > Status.NONE,
  );

  if (joinedParticipants.length < 2) {
    return false;
  }

  // every joined participant needs a name
  for (const joinedParticipant of joinedParticipants) {
    if (!joinedParticipant.participant) {
      return false;
    }
  }

  const payingParticipants = joinedParticipants.filter(
    (participant) => participant.status === Status.PAID && participant.amount,
  );
  if (!payingParticipants.length) {
    return false;
  }

  return true;
}

function validateDirect(data: TransactionFormState["direct"]): boolean {
  const from = data.from === NEW_PARTICIPANT_OPTION ? data.fromNew : data.from;
  const to = data.to === NEW_PARTICIPANT_OPTION ? data.toNew : data.to;
  return !!from && !!to && !!data.amount && from !== to;
}

export function validate(data: TransactionFormState): boolean {
  if (!data.description) {
    return false;
  }

  if (!data.date) {
    return false;
  }

  if (data.transactionType === TransactionType.DIRECT) {
    return validateDirect(data.direct);
  } else {
    return validateShared(data.shared);
  }
}
