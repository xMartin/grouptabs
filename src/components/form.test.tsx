import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import Form from "./form";
import {
  TransactionType,
  DocumentType,
  TransactionFormParticipantStatus as Status,
  TransactionFormParticipantInputType as InputType,
} from "../types";
import { createFormData } from "../util/transactionform";

let realDate: any;

beforeAll(() => {
  const currentDate = new Date("2020-02-20T20:20:20.202Z");
  realDate = Date;
  global.Date = class extends Date {
    constructor(date: any, ...args: any[]) {
      if (date) {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        return super(date, ...args);
      }

      return currentDate;
    }
  } as any;
});

afterAll(() => {
  global.Date = realDate;
});

it("renders empty form", () => {
  const tree = renderer
    .create(
      <Form
        mode="new"
        data={createFormData([])}
        onUpdateForm={vi.fn()}
        onUpdateSharedForm={vi.fn()}
        onUpdateDirectForm={vi.fn()}
        onUpdateParticipant={vi.fn()}
        onAddParticipant={vi.fn()}
        onSetAllJoined={vi.fn()}
        onSave={vi.fn()}
        onDelete={vi.fn()}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders prefilled form", () => {
  const tree = renderer
    .create(
      <Form
        mode="edit"
        data={createFormData(
          [
            {
              participant: "Martin",
              amount: -21,
            },
            {
              participant: "Jan",
              amount: 21,
            },
          ],
          {
            id: "123",
            type: DocumentType.TRANSACTION,
            tabId: "321",
            description: "DESCRIPTION",
            transactionType: TransactionType.SHARED,
            date: "2020-03-24",
            timestamp: "2020-03-24T20:23:21z",
            participants: [
              {
                participant: "Martin",
                amount: 5,
              },
              {
                participant: "Jan",
                amount: 0,
              },
            ],
          }
        )}
        onUpdateForm={vi.fn()}
        onUpdateSharedForm={vi.fn()}
        onUpdateDirectForm={vi.fn()}
        onUpdateParticipant={vi.fn()}
        onAddParticipant={vi.fn()}
        onSetAllJoined={vi.fn()}
        onSave={vi.fn()}
        onDelete={vi.fn()}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

describe('"all joined" button', () => {
  it("shows button for 2 unjoined participants", () => {
    render(
      <Form
        mode="new"
        data={{
          transactionType: TransactionType.SHARED,
          date: "2011-11-11",
          direct: {
            options: [],
          },
          shared: [
            {
              id: "1",
              inputType: InputType.NEW,
              status: Status.NONE,
            },
            {
              id: "2",
              inputType: InputType.NEW,
              status: Status.NONE,
            },
          ],
        }}
        onUpdateForm={vi.fn()}
        onUpdateSharedForm={vi.fn()}
        onUpdateDirectForm={vi.fn()}
        onUpdateParticipant={vi.fn()}
        onAddParticipant={vi.fn()}
        onSetAllJoined={vi.fn()}
        onSave={vi.fn()}
        onDelete={vi.fn()}
      />
    );

    expect(screen.getByText("all joined")).toBeInTheDocument();
  });

  it("shows no button for 2 joined and 1 unjoined participant", () => {
    render(
      <Form
        mode="new"
        data={{
          transactionType: TransactionType.SHARED,
          date: "2011-11-11",
          direct: {
            options: [],
          },
          shared: [
            {
              id: "1",
              inputType: InputType.NEW,
              status: Status.PAID,
            },
            {
              id: "2",
              inputType: InputType.NEW,
              status: Status.JOINED,
            },
            {
              id: "3",
              inputType: InputType.NEW,
              status: Status.NONE,
            },
          ],
        }}
        onUpdateForm={vi.fn()}
        onUpdateSharedForm={vi.fn()}
        onUpdateDirectForm={vi.fn()}
        onUpdateParticipant={vi.fn()}
        onAddParticipant={vi.fn()}
        onSetAllJoined={vi.fn()}
        onSave={vi.fn()}
        onDelete={vi.fn()}
      />
    );

    expect(screen.queryByText("all joined")).not.toBeInTheDocument();
  });
});
