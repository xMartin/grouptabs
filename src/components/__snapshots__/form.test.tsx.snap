// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders empty form 1`] = `
<form
  id="edit-entry-form"
  onSubmit={[Function]}
>
  <div
    className="form"
  >
    <div
      className="form-row"
    >
      <div
        className="form-row-input description"
      >
        <input
          onChange={[Function]}
          placeholder="Description"
          type="text"
          value=""
        />
      </div>
      <div
        className="form-row-input transaction-type"
      >
        <select
          onChange={[Function]}
          value="SHARED"
        >
          <option
            value="SHARED"
          >
            Shared
          </option>
          <option
            value="DIRECT"
          >
            Direct
          </option>
        </select>
      </div>
    </div>
    <div
      className="form-row"
    >
      <div
        className="form-row-input"
      >
        <div
          className="form-row-input date-input"
        >
          <button
            className="selected"
            onClick={[Function]}
            type="button"
          >
            today
          </button>
          <button
            className=""
            onClick={[Function]}
            type="button"
          >
            yesterday
          </button>
          <input
            onChange={[Function]}
            type="date"
            value="2020-02-20"
          />
        </div>
      </div>
    </div>
    <div
      style={
        Object {
          "display": "none",
        }
      }
    >
      <div
        className="direct-transaction"
      >
        <div
          className="form-row"
          style={
            Object {
              "display": "none",
            }
          }
        >
          <div
            className="form-row-input"
          >
            <select
              className="full-width"
              onChange={[Function]}
              value="NEW_PARTICIPANT"
            >
              <option
                value="NEW_PARTICIPANT"
              >
                New participant…
              </option>
            </select>
          </div>
        </div>
        <div
          className="form-row"
        >
          <div
            className="form-row-input"
          >
            <input
              autoFocus={true}
              onChange={[Function]}
              placeholder="Name …"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          className="direct-transaction-amount"
        >
          <svg
            height="16"
            width="16"
          >
            <path
              d="m15.511 8.5129c0-0.8974-1.0909-1.3404-1.7168-0.6973l-4.7832 4.7837v-11.573c0.019125-1.3523-2.0191-1.3523-2 0v11.572l-4.7832-4.7832c-0.94251-0.98163-2.3957 0.47155-1.4141 1.4141l6.49 6.4911c0.3878 0.387 1.0228 0.391 1.414 0l6.4903-6.4906c0.1935-0.1883 0.30268-0.4468 0.3027-0.7168z"
            />
          </svg>
          <input
            onChange={[Function]}
            placeholder="0"
            step="any"
            type="number"
            value=""
          />
        </div>
        <div
          className="form-row"
          style={
            Object {
              "display": "none",
            }
          }
        >
          <div
            className="form-row-input"
          >
            <select
              className="full-width"
              onChange={[Function]}
              value="NEW_PARTICIPANT"
            >
              <option
                value="NEW_PARTICIPANT"
              >
                New participant…
              </option>
            </select>
          </div>
        </div>
        <div
          className="form-row"
        >
          <div
            className="form-row-input"
          >
            <input
              autoFocus={true}
              onChange={[Function]}
              placeholder="Name …"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      style={
        Object {
          "display": "",
        }
      }
    >
      <div
        className="form-row"
      >
        <div
          className="form-row-input"
        >
          <div
            className="newParticipantInput selected"
          >
            <span
              className="participant"
            >
              <input
                autoFocus={false}
                onChange={[Function]}
                placeholder="Name …"
                type="text"
                value=""
              />
            </span>
            <span
              className="participationStatus"
            >
              <span
                className="joinedButtonWrapper"
              >
                <button
                  className="selected"
                  onClick={[Function]}
                  type="button"
                >
                  joined
                </button>
              </span>
              <button
                className="paid-button"
                onClick={[Function]}
                type="button"
              >
                paid
              </button>
            </span>
          </div>
          <div
            className="newParticipantInput selected"
          >
            <span
              className="participant"
            >
              <input
                autoFocus={false}
                onChange={[Function]}
                placeholder="Name …"
                type="text"
                value=""
              />
            </span>
            <span
              className="participationStatus"
            >
              <span
                className="joinedButtonWrapper"
              >
                <button
                  className="selected"
                  onClick={[Function]}
                  type="button"
                >
                  joined
                </button>
              </span>
              <button
                className="paid-button"
                onClick={[Function]}
                type="button"
              >
                paid
              </button>
            </span>
          </div>
        </div>
      </div>
      <div
        className="form-row"
      >
        <div
          className="form-row-input"
        >
          <button
            onClick={[MockFunction]}
            type="button"
          >
            + new participant
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    className="row"
  />
</form>
`;

exports[`renders prefilled form 1`] = `
<form
  id="edit-entry-form"
  onSubmit={[Function]}
>
  <div
    className="form"
  >
    <div
      className="form-row"
    >
      <div
        className="form-row-input description"
      >
        <input
          onChange={[Function]}
          placeholder="Description"
          type="text"
          value="DESCRIPTION"
        />
      </div>
      <div
        className="form-row-input transaction-type"
      >
        <select
          onChange={[Function]}
          value="SHARED"
        >
          <option
            value="SHARED"
          >
            Shared
          </option>
          <option
            value="DIRECT"
          >
            Direct
          </option>
        </select>
      </div>
    </div>
    <div
      className="form-row"
    >
      <div
        className="form-row-input"
      >
        <div
          className="form-row-input date-input"
        >
          <button
            className=""
            onClick={[Function]}
            type="button"
          >
            today
          </button>
          <button
            className=""
            onClick={[Function]}
            type="button"
          >
            yesterday
          </button>
          <input
            onChange={[Function]}
            type="date"
            value="2020-03-24"
          />
        </div>
      </div>
    </div>
    <div
      style={
        Object {
          "display": "none",
        }
      }
    >
      <div
        className="direct-transaction"
      >
        <div
          className="form-row"
        >
          <div
            className="form-row-input"
          >
            <select
              className="full-width"
              onChange={[Function]}
              value="Martin"
            >
              <option
                value="Martin"
              >
                Martin
              </option>
              <option
                value="Jan"
              >
                Jan
              </option>
              <option
                value="NEW_PARTICIPANT"
              >
                New participant…
              </option>
            </select>
          </div>
        </div>
        <div
          className="direct-transaction-amount"
        >
          <svg
            height="16"
            width="16"
          >
            <path
              d="m15.511 8.5129c0-0.8974-1.0909-1.3404-1.7168-0.6973l-4.7832 4.7837v-11.573c0.019125-1.3523-2.0191-1.3523-2 0v11.572l-4.7832-4.7832c-0.94251-0.98163-2.3957 0.47155-1.4141 1.4141l6.49 6.4911c0.3878 0.387 1.0228 0.391 1.414 0l6.4903-6.4906c0.1935-0.1883 0.30268-0.4468 0.3027-0.7168z"
            />
          </svg>
          <input
            onChange={[Function]}
            placeholder="0"
            step="any"
            type="number"
            value=""
          />
        </div>
        <div
          className="form-row"
        >
          <div
            className="form-row-input"
          >
            <select
              className="full-width"
              onChange={[Function]}
              value="Jan"
            >
              <option
                value="Martin"
              >
                Martin
              </option>
              <option
                value="Jan"
              >
                Jan
              </option>
              <option
                value="NEW_PARTICIPANT"
              >
                New participant…
              </option>
            </select>
          </div>
        </div>
      </div>
    </div>
    <div
      style={
        Object {
          "display": "",
        }
      }
    >
      <div
        className="form-row"
      >
        <div
          className="form-row-input"
        >
          <div
            className="participantInput selected paid"
          >
            <span
              className="participant"
            >
              Martin
            </span>
            <span
              className="participationStatus"
            >
              <button
                className="paid-button selected"
                onClick={[Function]}
                type="button"
              >
                paid
              </button>
              <span
                className="amountInput"
              >
                <input
                  autoFocus={false}
                  onChange={[Function]}
                  placeholder="0"
                  step="any"
                  type="number"
                  value={5}
                />
              </span>
            </span>
          </div>
          <div
            className="participantInput selected"
          >
            <span
              className="participant"
            >
              Jan
            </span>
            <span
              className="participationStatus"
            >
              <span
                className="joinedButtonWrapper"
              >
                <button
                  className="selected"
                  onClick={[Function]}
                  type="button"
                >
                  joined
                </button>
              </span>
              <button
                className="paid-button"
                onClick={[Function]}
                type="button"
              >
                paid
              </button>
            </span>
          </div>
        </div>
      </div>
      <div
        className="form-row"
      >
        <div
          className="form-row-input"
        >
          <button
            onClick={[MockFunction]}
            type="button"
          >
            + new participant
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    className="row button-row"
  >
    <span
      className="fake-link delete"
      onClick={[Function]}
    >
      Delete transaction
    </span>
  </div>
</form>
`;
